smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructuredMemberWriter.java

```diff
@@ -65,6 +69,130 @@ void writeMembers(TypeScriptWriter writer, Shape shape) {
+            writer.write("$T.filterSensitiveLog", symbolProvider.toSymbol(collectionMember));
+        } else if (memberShape instanceof CollectionShape) {
+            // Iterate over array items, and call array specific function on each member.
+            writer.openBlock("item => item.map(", ")",
+                () -> {
+                    MemberShape nestedCollectionMember = ((CollectionShape) memberShape).getMember();
+                    writeCollectionFilterSensitiveLog(writer, nestedCollectionMember);
+                }
+            );
+        } else if (memberShape instanceof MapShape) {
+            // Iterate over Object entries, and call reduce to repopulate map.
+            writer.openBlock("item => Object.entries(item).reduce(", ")",
+                () -> {
+                    MemberShape mapMember = ((MapShape) memberShape).getValue();
+                    writeMapFilterSensitiveLog(writer, mapMember);
+                }
+            );
+        } else {
+            // This path will never reach because of recursive isIterationRequired.
+            // Adding it to not break the code, if it does reach in future.
+            writer.write("item => item");
+        }
+    }
+
+    /**
+     * Recursively writes filterSensitiveLog for MapShape.
+     */
+    void writeMapFilterSensitiveLog(TypeScriptWriter writer, MemberShape mapMember) {
+        // Reducer is common to all shapes.
+        writer.openBlock("(acc: any, [key, value]: [string, $T]) => ({", "}), {}",
+            symbolProvider.toSymbol(mapMember),
+            () -> {
+                writer.write("...acc,");
+                Shape memberShape = model.expectShape(mapMember.getTarget());
+                if (memberShape instanceof StructureShape) {
+                    // Call filterSensitiveLog on Structure.
+                    writer.write("[key]: $T.filterSensitiveLog(value),",
+                        symbolProvider.toSymbol(mapMember));
+                } else if (memberShape instanceof CollectionShape) {
+                    writer.openBlock("[key]: value.map(", "),",
+                        () -> {
+                            MemberShape collectionMember = ((CollectionShape) memberShape).getMember();
+                            writeCollectionFilterSensitiveLog(writer, collectionMember);
+                        }
+                    );
+                } else if (memberShape instanceof MapShape) {
+                    writer.openBlock("[key]: Object.entries(value).reduce(", "),",
+                        () -> {
+                            MemberShape nestedMapMember = ((MapShape) memberShape).getValue();
+                            writeMapFilterSensitiveLog(writer, nestedMapMember);
+                        }
+                    );
+                } else {
+                    // This path will never reach because of recursive isIterationRequired.
+                    // Adding it to not break the code, if it does reach in future.
```

Comment:
Should we be throwing an error here?

Reply:
Makes sense! Added in commit https://github.com/awslabs/smithy-typescript/pull/170/commits/e00a880ef6b8562e55b42efe967d03608cf12f93

Verified that codegen will throw error for this case, by return true in isIterationRequired:
```console
Projection accessanalyzer.2019-11-01 failed: software.amazon.smithy.codegen.core.CodegenException: CollectionFilterSensitiveLog attempted for string while it was not required
software.amazon.smithy.typescript.codegen.StructuredMemberWriter.writeCollectionFilterSensitiveLog(StructuredMemberWriter.java:99)
software.amazon.smithy.typescript.codegen.StructuredMemberWriter.lambda$writeFilterSensitiveLog$5(StructuredMemberWriter.java:164)
software.amazon.smithy.utils.CodeWriter.openBlock(CodeWriter.java:909)
software.amazon.smithy.utils.CodeWriter.openBlock(CodeWriter.java:819)
software.amazon.smithy.typescript.codegen.StructuredMemberWriter.writeFilterSensitiveLog(StructuredMemberWriter.java:162)
software.amazon.smithy.typescript.codegen.StructureGenerator.lambda$null$2(StructureGenerator.java:180)
```